### DISCUSSION TITLE
My custom network driver not considered by zephyr as a network driver (No SOURCES given to Zephyr library: drivers__ethernet)

### DISCUSSION THREAD

**Discussion by ofirshe** [zephyrproject-rtos/zephyr#53210] [Q&A] [answered]
Hey, 

I developed a custom network driver for our board which consists NXP RT1062 microcontroller. Our driver based on eth_mcux.c network driver with many customizations. The driver is being built as a zephyr module and being appended in the CMakeLists.txt file to ZEPHYR_EXTRA_MODULES. However, the driver is not considered as a network driver by zephyr and I get the warning message once compiling the project: 

west build -b xsightlabs_evb_a -- -DE1_STUB=1
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base (cached)).
-- Application: C:/Users/ofirs/epu-bsp
-- Cache files will be written to: C:/Users/ofirs/epu-bsp/zephyr/.cache
-- Zephyr version: 3.2.99 (C:/Users/ofirs/epu-bsp/zephyr)
-- Found west (found suitable version "0.13.1", minimum required is "0.7.1")
-- Board: xsightlabs_evb_a
-- Found toolchain: gnuarmemb (C:\Users\ofirs\epu-bsp\gcc-arm-none-eabi-10.3-2021.10)
-- Found BOARD.dts: C:/Users/ofirs/epu-bsp/boards/arm/xsightlabs_evb/xsightlabs_evb_a.dts
node '/soc/i2c@403f8000/fs1403@8' compatible 'tdk,fs140x' has unknown vendor prefix 'tdk'
-- Generated zephyr.dts: C:/Users/ofirs/epu-bsp/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: C:/Users/ofirs/epu-bsp/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: C:/Users/ofirs/epu-bsp/build/zephyr/dts.cmake
Parsing C:/Users/ofirs/epu-bsp/zephyr/Kconfig
Loaded configuration 'C:/Users/ofirs/epu-bsp/build/zephyr/.config'
No change to configuration in 'C:/Users/ofirs/epu-bsp/build/zephyr/.config'
No change to Kconfig header in 'C:/Users/ofirs/epu-bsp/build/zephyr/include/generated/autoconf.h'
Load components for MIMXRT1062:
driver_common component is included.
driver_reset component is included.
device_CMSIS component is included.
CMSIS_Include_core_cm component is included.
device_system component is included.
driver_trng component is included.
driver_lpi2c component is included.
driver_lpspi component is included.
driver_lpuart component is included.
driver_igpio component is included.
driver_flexspi component is included.
driver_cache_armv7_m7 component is included.
CMake Warning at zephyr/CMakeLists.txt:824 (message):
  No SOURCES given to Zephyr library: drivers__ethernet

  Excluding target from build.

-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/ofirs/epu-bsp/build
-- west build: building application
[9/17] Linking C executable zephyr\zephyr_pre0.elf

[12/17] Linking C executable zephyr\zephyr_pre1.elf

[17/17] Linking C executable zephyr\zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      133596 B         8 MB      1.59%
             RAM:      129808 B       768 KB     16.51%
            ITCM:        3520 B       128 KB      2.69%
            DTCM:       13908 B       128 KB     10.61%
        IDT_LIST:          0 GB         2 KB      0.00%

I will be grateful to get your assist with that. @DerekSnell @dbaluta 

Thanks in advance.

**Comment by ofirshe** [comment]
![image](https://user-images.githubusercontent.com/37531225/208700056-73461b19-e8e2-415b-a5ea-0ccbdbb29181.png)
I see that also no libdrivers_ethernet.a created in build/zephyr/drivers/ethernet folder

**Reply by ofirshe** [reply]
added my driver in the zephyr directory at drivers/ethernet folder.

**Comment by SavchenkoIvan** [comment]
Hello! I faced exactly the same problem and don't know how to solve it. Did you manage to fix it?

**Reply by ofirshe** [reply]
Yes, 

the only way to fix that is to add your driver in the zephyr directory in drivers/ethernet folder.

**Comment by yashi** [comment]
Hi, what exactly is your problem? 

`drivers__ethernet` is just a `.a` archive file, which combines all object files _under_ `drivers/ethernet`.  Your custom driver is not under the directory and won't be in the archive for sure.  But, that doesn't and shouldn't make any difference at all.  It doesn't mean the build system doesn't consider yours not an Ethernet driver.  In fact, the build system doesn't care what is what. It just builds selected source files and generates archives under each dir just because that makes linking easier.

Zephyr proper doesn't care either.  What matter is that either your driver implements the required API or not.   And subsystems can call the APIs when they need to.

Hope this helps.

**Reply by ofirshe** [reply]
Enabling CONFIG_NETWORKING in project configuration leads to zephyr/drivers/ethernet library to be enabled. However, there are no sources given to this zephyr library because we are using our own custom network drivers which is not located in this subdirectory, in this results in printing this warning while building the project : 

No SOURCES given to Zephyr library: drivers__ethernet

Excluding target from build.

I just wanted to solve that issue and no way to solve that except putting my files in this directory that's it.

**Reply by yashi** [reply]
The following command emulates what you see.
```
west build -b mimxrt1060_evk zephyr/samples/hello_world -- -DCONFIG_NETWORKING=y -DCONFIG_ETH_MCUX=n
    :
CMake Warning at /home/yashi/src/zephyr-project/zephyr/CMakeLists.txt:824 (message):
  No SOURCES given to Zephyr library: drivers__ethernet

  Excluding target from build.
```

If you don't want to see it you can disable Ethernet Driver:
```
west build -b mimxrt1060_evk zephyr/samples/hello_world -- -DCONFIG_NETWORKING=y -DCONFIG_ETH_MCUX=n -DCONFIG_NET_L2_ETHERNET=n
```

**Reply by ofirshe** [reply]
but I don't want to disable it. Why should I disable it ?

**Reply by yashi** [reply]
Because other parts of Zephyr depends on `CONFIG_NET_L2_ETHERNET`? Hmm.  You are right. In that case my assumption is wrong.  Please see the other answer.

**Comment by yashi** [comment, marked as answer]
I found another solution.  There is a macro called `zephyr_library_amend()` in [`zephyr/cmake/modules/extensions.cmake`](https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v3.2.0/cmake/modules/extensions.cmake#L425-L463)  Does it help?

```cmake
# Provides amend functionality to a Zephyr library for out-of-tree usage.
#
# When called from a Zephyr module, the corresponding zephyr library defined
# within Zephyr will be looked up.
#
# Note, in order to ensure correct library when amending, the folder structure in the
# Zephyr module must resemble the structure used in Zephyr, as example:
#
# Example: to amend the zephyr library created in
# ZEPHYR_BASE/drivers/entropy/CMakeLists.txt
# add the following file:
# ZEPHYR_MODULE/drivers/entropy/CMakeLists.txt
# with content:
# zephyr_library_amend()
# zephyr_library_add_sources(...)
#
# It is also possible to use generator expression when amending to Zephyr
# libraries.
#
# For example, in case it is required to expose the Zephyr library's folder as
# include path then the following is possible:
# zephyr_library_amend()
# zephyr_library_include_directories($<TARGET_PROPERTY:SOURCE_DIR>)
#
# See the CMake documentation for more target properties or generator
# expressions.
#
macro(zephyr_library_amend)
  # This is a macro because we need to ensure the ZEPHYR_CURRENT_LIBRARY and
  # following zephyr_library_* calls are executed within the scope of the
  # caller.
  if(NOT ZEPHYR_CURRENT_MODULE_DIR)
    message(FATAL_ERROR "Function only available for Zephyr modules.")
  endif()

  zephyr_library_get_current_dir_lib_name(${ZEPHYR_CURRENT_MODULE_DIR} lib_name)

  set(ZEPHYR_CURRENT_LIBRARY ${lib_name})
endmacro()
```

**Reply by ofirshe** [reply]
Hey, 

I've already chosen the option of creating a local copy of zephyr and insert my files inside ethernet directory but I believe your solution will work. 

Thank you for this suggestion
